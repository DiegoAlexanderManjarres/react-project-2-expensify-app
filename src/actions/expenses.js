import uuidv4 from 'uuid';
import database from '../firebase/firebase'

// ADD_EXENSE
export const addExpense = (expense) => ({
   type: 'ADD_EXPENSE',
   expense
})

export const startAddExpense = (expenseData = {}) => {
	return (dispatch) => {

		const { 
			description = '', 
			note = '', 
			amount = 0, 
			createdAt = 0
		} = expenseData // destructured obj

		const expense = { description, note, amount, createdAt } 

		return database.ref('expenses').push(expense).then(ref => {
			dispatch(addExpense({
				id: ref.key,
				...expense
			}))
		})  
	}
} 

// REMOVE_ EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE', 
  id 
})

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

// SET_EXPENSES
export const setExpenses = (expenses) => ({
	type: 'SET_EXPENSES',
	expenses
})

//export const startSetExpenses

export const startSetExpenses = () => (dispatch) => (
	 database.ref('expenses').once('value').then(snapshot => {
		 const expenses = []
			snapshot.forEach(childSnapshots => {
				expenses.push({ id: childSnapshots.key, ...childSnapshots.val() })
			})
			dispatch(setExpenses(expenses))				
		})
	)

export const startRemoveExpense = ({ id }) => (dispatch) => (
	database.ref(`expenses/${id}`).set(null)
		.then(() => dispatch(removeExpense({ id })))
)

export const startEditExpense = (id, expenseUpdates) => (dispatch) => ( 
	database.ref(`expenses/${id}`).update(expenseUpdates)
		.then(() => dispatch(editExpense(id, expenseUpdates)))
)		

	
 

	
		 		
 
