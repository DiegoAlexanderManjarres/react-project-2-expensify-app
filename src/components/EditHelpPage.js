import React from 'react';

const EditHelpPage = () => (
  <div>My edit help page</div>
)

export default EditHelpPage;