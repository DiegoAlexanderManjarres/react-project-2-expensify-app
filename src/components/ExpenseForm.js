import React from 'react'
import Parser from 'html-react-parser'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'

export default class ExpenseForm extends React.Component {
 	constructor(props) {
		super(props)
		this.state = {
			description: props.expense ? props.expense.description : '',
			note: props.expense ? props.expense.note : '',
			amount: props.expense ? (props.expense.amount / 100).toString() : '',
			createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
			calendarFocused: false,
			error: '',
			titleButton: props.expense? 'Save Expense' : 'Add Expense'
		}
  }
  
   onDescriptionChange = (e) => {
		e.preventDefault();
		const description = e.target.value;
		this.setState(() => ({ description }));
  }
  	onNoteChange = (e) => {
		e.preventDefault()  
		const note = e.target.value;
		this.setState(() => ({ note }))
  }
 	onAmountChange = (e) => {
		e.preventDefault();
		const reg = /^\d{1,}(\.\d{0,2})?$/;
		const amount = e.target.value;
		if (!amount || amount.match(reg)) {
			this.setState(() => ({ amount }));
		} 
  }
  onDateChange = (createdAt) => {
		if (createdAt) {
			this.setState(() => ({ createdAt }));
		}    
   }
  	onFocusChange = ({ focused }) => {
		this.setState(() => ({ calendarFocused: focused }))
   }
   onSubmit = (e) => {
		e.preventDefault();
		if (!this.state.description || !this.state.amount) {
		// Set error state to 'Please provide description and amount'
		// used module html-react-parser for the template string with html
			this.setState(() => ({
				...this.state, 
				error: `
				   Please provide 
					<strong> description </strong> 
					and 
					<strong> amount </strong>.
				`
			}));
		} else {
		// Clear the error
			this.setState(() => ({ ...this.state, error: '' }));
			this.props.onSubmit({
				description: this.state.description,
				amount: parseFloat(this.state.amount, 10) * 100,
				createdAt: this.state.createdAt.valueOf(),
				note: this.state.note
			})
		}
   }
   render(){
		return (	
			<form className="form" onSubmit={this.onSubmit}>
				{
					this.state.error && 
					<p className="form__error">{Parser(this.state.error)}</p>
				}
					<input
						type='text'
						placeholder='Description'
						autoFocus
						className="text-input"
						value={this.state.description}
						onChange={this.onDescriptionChange}
					/>
					<input 
						type="text"
						placeholder="Amount"
						className="text-input"
						value={this.state.amount}
						onChange={this.onAmountChange}            
					/>
					<SingleDatePicker 
						date={this.state.createdAt}
						onDateChange={this.onDateChange}
						focused={this.state.calendarFocused}
						onFocusChange={this.onFocusChange}
						numberOfMonths={1}
						isOutsideRange={() => false}
						block
						/>
					<textarea 
						placeholder="Add a note for your expense (optional)"
						value={this.state.note}
						onChange={this.onNoteChange}
						className="textarea"/>
					<div>
						<button className="button">{this.state.titleButton}</button>
					</div>	
			</form>
			
		)
  }
}
