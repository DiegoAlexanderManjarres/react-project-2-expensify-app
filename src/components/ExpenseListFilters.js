import React from 'react'
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates'
import 'react-dates/initialize'
import uuidv4 from 'uuidv4'
import { 
	setTextFilter, 
	sortByAmount, 
	sortByDate, 
	setStartDate,
	setEndDate 
} from '../actions/filters'


export class  ExpenseListFilters extends React.Component {
	state = {
		calendarFocused: null,
		idGennerator: uuidv4()
	}  
	onDatesChange = ({ startDate, endDate }) => {
		this.props.setStartDate(startDate)
		this.props.setEndDate(endDate)
	}
	onFocusChange = (calendarFocused) => (
		this.setState(() => ({ calendarFocused }))
	)
	onTextChange = (e) => {
		const text = e.target.value;
		this.props.setTextFilter(text);
	}
	onSortChange = (e) => (
		e.target.value === 'date' 
		? this.props.sortByDate() 
		: this.props.sortByAmount()        
	)		
	render() {
		return (
			<div className="content-container">
				<div className="input-group">
					<div className="input-group__item">
						<input 
							type='text' 
							value={this.props.filters.text}
							onChange={this.onTextChange}
							className="select"
							placeholder="Search Expenses"/>							
					</div>
					<div className="input-group__item">
						<select
							onChange={this.onSortChange}
							className="text-input">
							<option value="date">Date</option>
							<option value="amount">Amount</option>
						</select>
					</div>
					<div className="input-group__item">
						<DateRangePicker 
							startDate={this.props.filters.startDate}
							endDate={this.props.filters.endDate}
							onDatesChange={this.onDatesChange}
							focusedInput={this.state.calendarFocused}
							onFocusChange={this.onFocusChange}
							numberOfMonths={1}
							isOutsideRange={() => false}
							showClearDates={true}
							startDateId={this.state.idGennerator}
							endDateId={this.state.idGennerator}
						/>
					</div>
				</div>
			</div>
		)
	}
}

const mapDispatchToProps = (dispatch) => ({
	setTextFilter: (text) => dispatch(setTextFilter(text)),
	sortByDate: () => dispatch(sortByDate()),
	sortByAmount: (amount) => dispatch(sortByAmount(amount)),
	setStartDate: (startDate) => dispatch(setStartDate(startDate)),
	setEndDate: (endDate) => dispatch(setEndDate(endDate))
})

const mapStateToProps = (state) => ({
	filters: state.filters
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)