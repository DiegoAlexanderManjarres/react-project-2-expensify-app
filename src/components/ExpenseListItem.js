// Export a Stateless functional component
// description, amount, createdAt
import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import moment from 'moment'
import numeral from 'numeral'

export const ExpenseListItem = ({id, description, amount, createdAt}) => {
   return (    
	  	<li  data-id={id} className="list-item-wrapper">
			<Link className="list-item"
				to={`/edit/${id}`}>
				<div>
					<h3 className="list-item__title">{description}</h3>
					<strong className="list-item__sub-title">
						{moment(createdAt).format('MMMM Do, YYYY')}
					</strong>
				</div>
				<h3 className="list-item__data">
					{numeral(amount / 100).format('$0,0.00')}
				</h3>				
			</Link>			       
	  	</li>   
   )
}
export default connect()(ExpenseListItem);