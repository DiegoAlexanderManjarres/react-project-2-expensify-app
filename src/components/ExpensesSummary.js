import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import selectExpenseTotal from '../selectors/Exenses-total'
import selectExpenses from '../selectors/expenses'
import numeral from 'numeral'

export const ExpensesSummary = ({ expenses }) => {
   const total = numeral(selectExpenseTotal(expenses) / 100).format('$0,0.00')
   const expensesWord = expenses.length === 1 ? 'expense' : 'expenses'
   return (            
      <div className="page-header">
         <div className="content-container">
            { expenses.length ? (
               <h1 className="page-header__title">Viewing
                  <strong> {expenses.length} </strong> 
                  {expensesWord} totalling
                  <strong> {total} </strong>
               </h1>
            ) : (
               <h1 className="page-header__title">Add expense</h1>               
            )} 
            <div className="page-header__actions">
               <Link className="button" to="/create">Add Expense</Link>
            </div> 
         </div> 
          
      </div>       
   )
} 

const mapStateToProps = (state) => ({
	expenses: selectExpenses(state.expenses, state.filters)
})

export default connect(mapStateToProps)(ExpensesSummary);
