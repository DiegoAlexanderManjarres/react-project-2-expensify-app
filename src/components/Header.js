import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
  <header>
    <h1>Expensify</h1>
    <div>
      <NavLink to="/" activeClassName="is-active" exact={true}>
        <strong> Dashboard </strong> 
      </NavLink>
      <br />
      <NavLink to="/create" activeClassName="is-active"> 
        <strong> Create-Expense </strong> 
      </NavLink>
    </div>
  </header>
)

export default Header;