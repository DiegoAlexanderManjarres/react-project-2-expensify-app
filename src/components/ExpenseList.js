import React from "react";
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';
// connect for components that needs to read from the store or didpatch actions

export const ExpenseList = (props) => (
	<div className="content-container">
		<div className="list-header">
			<div className="show-for-mobile">Expenses</div>
			<div className="show-for-desktop">Expense</div>
			<div className="show-for-desktop">Amount</div>
		</div>
		<div className="list-body">
			{
				!props.expenses.length ? (
					<div className="list-item-wrapper list-item list-item--message">
						<span>No expenses</span>
					</div>
					) : (
					<ul>    
						{
							props.expenses.map(expense => (
								<ExpenseListItem 
									key={expense.id} 
									{...expense} 
									{...props}/>
							))
						}
					</ul>    
				)
			}
		</div>	
	</div>
);

const mapStateToProps = (state) => ({
	expenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);