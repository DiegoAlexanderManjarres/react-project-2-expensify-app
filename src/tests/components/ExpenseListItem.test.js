import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseListItem } from '../../components/ExpenseListItem'
import expenses from '../fixtures/expenses'

// Render ExpenseListItem with fixture data
test('should render expenseListItem with fixture date', () => {
   const wrapper = shallow(<ExpenseListItem { ...expenses[1] }/>)
   expect(wrapper).toMatchSnapshot()
})