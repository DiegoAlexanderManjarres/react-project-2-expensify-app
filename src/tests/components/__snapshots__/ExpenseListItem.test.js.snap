// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expenseListItem with fixture date 1`] = `
<li
  className="list-item-wrapper"
  data-id="2"
>
  <Link
    className="list-item"
    replace={false}
    to="/edit/2"
  >
    <div>
      <h3
        className="list-item__title"
      >
        rent
      </h3>
      <strong
        className="list-item__sub-title"
      >
        December 26th, 1969
      </strong>
    </div>
    <h3
      className="list-item__data"
    >
      $1,095.00
    </h3>
  </Link>
</li>
`;
